green_cli/authenticators/__init__.py

```diff
@@ -0,0 +1,170 @@
+        return gdk.register_user(session, self.hw_device, self.mnemonic)
+
+
+class ConfigProperty:
+    """A piece of data that is stored in a file in the config directory"""
+
+    def __init__(self, config_dir, filename, prompt_fn, file_perms=stat.S_IRUSR | stat.S_IWUSR):
+        self.filename = os.path.join(config_dir, filename)
+        self.prompt_fn = prompt_fn
+        self.file_perms = file_perms
+
+    def get(self):
+        """Read the value from the config file, or failing that prompt the user"""
+        try:
+            return open(self.filename).read()
+        except IOError:
+            value = self.prompt_fn()
+            self.set(value)
+            return value
+
+    def set(self, value):
+        open(self.filename, 'w').write(value)
+        os.chmod(self.filename, self.file_perms)
+
+
+class MnemonicOnDisk:
+    """Persist a mnemonic using the filesystem"""
+
+    def __init__(self, config_dir):
+        # mnemonic file has read-only permissions to prevent accidental deletion
+        prompt_fn = lambda: getpass('Mnemonic: ')
+        self.mnemonic_prop = ConfigProperty(config_dir, 'mnemonic', prompt_fn, stat.S_IRUSR)
+
+    @property
+    def _mnemonic(self):
+        return self.mnemonic_prop.get()
+
+    @_mnemonic.setter
+    def _mnemonic(self, mnemonic):
+        """Write mnemonic to config file"""
+        try:
+            self.mnemonic_prop.set(mnemonic)
+        except PermissionError:
+            message = (
+                "Refusing to overwrite mnemonic file {}\n"
+                "First backup and then delete or change file permissions"
+                .format(self.mnemonic_prop.filename))
+            raise click.ClickException(message)
+
+
+class SoftwareAuthenticator(Authenticator, MnemonicOnDisk):
+    """Represent a 'software signer' which passes the mnemonic to the gdk for authentication
+    """
+
+    @property
+    def hw_device(self):
+        return json.dumps({})
+
+    @property
+    def mnemonic(self):
+        return self._mnemonic
+
+    @property
+    def password(self):
+        return ''
+
+    def create(self, session_obj):
+        """Create and register a new wallet"""
+        self._mnemonic = gdk.generate_mnemonic()
+        return self.register(session_obj)
+
+    def set_mnemonic(self, mnemonic):
+        mnemonic = ' '.join(mnemonic.split())
+        logging.debug("mnemonic: '{}'".format(mnemonic))
+        if not gdk.validate_mnemonic(mnemonic):
+            raise click.ClickException("Invalid mnemonic")
+        self._mnemonic = mnemonic
+
+
+class HardwareDevice(Authenticator):
+    """Represents what the gdk refers to as a 'hardware device'.
+
+    Not necessarily an actual hardware device, but anything that implements the required hardware
+    device interface
+    """
+
+    @property
+    def hw_device(self):
+        return json.dumps({'device': {'name': self.name, 'supports_liquid': 1,
+            'supports_low_r': True, 'supports_arbitrary_scripts': True}})
```

Comment:
Note sure we should default supports_low_r' to True or 'supports_arbitrary_scripts' to True. I don't think any currently released HW supports low-r sigs and Trezor doesn't support arbitrary scripts IIRC.